File: backend/admin-service/src/services/orderContract.test.ts
import {
  normalizeServiceCategory,
  toCanonicalOrderContract,
  withCanonicalOrderContract,
} from './orderContract';

describe('canonical order contract', () => {
  it('maps legacy service values without replacing the detail fields', () => {
    expect(normalizeServiceCategory({ service_category: 'food_delivery', service_sub_type: 'food_delivery' })).toBe('food');
    expect(normalizeServiceCategory({ service_sub_type: 'towing_mobil' })).toBe('towing');
    expect(normalizeServiceCategory({ model: 'aggregator', logistics_provider: 'jne' })).toBe('aggregator');
    expect(normalizeServiceCategory({ model: 'unknown_future_model' })).toBeNull();
  });

  it('builds typed metadata only from persisted order facts', () => {
    const contract = toCanonicalOrderContract({
      id: 'order-1',
      customer_id: 'customer-1',
      model: 'aggregator',
      service_code: 'jne_reg',
      logistics_provider: 'jne',
      logistics_service_type: 'REG',
      logistics_tariff_idr: 18000,
      logistics_net_cost_idr: 14000,
      status: 'pending_assignment',
      total_price_idr: 18000,
      state_version: 3,
    });

    expect(contract.service.category).toBe('aggregator');
    expect(contract.service.service_code).toBe('jne_reg');
    expect(contract.service.metadata.aggregator).toEqual({
      provider: 'jne',
      service_type: 'REG',
      tariff_idr: 18000,
      net_cost_idr: 14000,
      awb_number: null,
    });
    expect(contract.order_state.state_version).toBe(3);
  });

  it('keeps unknown services degraded-safe and preserves legacy response fields', () => {
    const result = withCanonicalOrderContract({ id: 'order-2', model: 'future_service', status: 'unknown' });
    expect(result.service_category).toBeNull();
    expect(result.order_contract.service.degraded).toBe(true);
    expect(result.order_contract.service.metadata).toEqual({});
    expect(result.id).toBe('order-2');
  });
});

File: backend/order-service/internal/domain/order_contract_test.go
package domain

import "testing"

func TestCanonicalServiceCategoryForDoesNotInventUnknownCategories(t *testing.T) {
	tests := []struct {
		name     string
		order    Order
		expected *CanonicalServiceCategory
	}{
		{
			name:     "legacy food subtype",
			order:    Order{ServiceSubType: "food_delivery"},
			expected: canonicalCategoryPtr(CanonicalFood),
		},
		{
			name:     "aggregator provider",
			order:    Order{LogisticsProvider: "jne"},
			expected: canonicalCategoryPtr(CanonicalAggregator),
		},
		{
			name:     "unknown model",
			order:    Order{Model: "future_service"},
			expected: nil,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			actual := CanonicalServiceCategoryFor(&tt.order)
			if (actual == nil) != (tt.expected == nil) {
				t.Fatalf("category = %v, expected %v", actual, tt.expected)
			}
			if actual != nil && *actual != *tt.expected {
				t.Fatalf("category = %q, expected %q", *actual, *tt.expected)
			}
		})
	}
}

func TestApplyCanonicalOrderContractPreservesTypedFacts(t *testing.T) {
	order := &Order{
		ID:                   "order-1",
		CustomerID:           "customer-1",
		Model:                "aggregator",
		LogisticsProvider:    "jne",
		LogisticsServiceType: "REG",
		LogisticsTariffIDR:   12000,
		LogisticsNetCostIDR:  10000,
		StateVersion:         0,
	}

	order.ApplyCanonicalOrderContract()

	if order.ServiceCategory != CanonicalAggregator {
		t.Fatalf("service category = %q", order.ServiceCategory)
	}
	if order.ContractVersion != CurrentOrderContractVersion {
		t.Fatalf("contract version = %q", order.ContractVersion)
	}
	if order.StateVersion != 1 {
		t.Fatalf("state version = %d", order.StateVersion)
	}
	if order.ServiceMetadata.Aggregator == nil || order.ServiceMetadata.Aggregator.Provider != "jne" {
		t.Fatalf("aggregator metadata was not preserved: %+v", order.ServiceMetadata.Aggregator)
	}
}

func TestBuildOrderServiceMetadataKeepsRoadsideFactsStructured(t *testing.T) {
	condition := "flat"
	order := &Order{
		ServiceSubType: "tambal_ban_tubeless",
		TambalBanReport: &TambalBanReport{
			TireConditionBefore: &condition,
		},
	}

	metadata := BuildOrderServiceMetadata(order)
	if metadata.Roadside == nil {
		t.Fatal("roadside metadata is nil")
	}
	if metadata.Roadside.VehicleFacts["tire_condition_before"] != "flat" {
		t.Fatalf("roadside vehicle facts = %+v", metadata.Roadside.VehicleFacts)
	}
}

func canonicalCategoryPtr(category CanonicalServiceCategory) *CanonicalServiceCategory {
	return &category
}
